tests/test_restriction.cpp

```diff
@@ -0,0 +1,288 @@
+    gyro::icntl[Param::ntheta_exp]     = (std::get<0>(GetParam()) % 3) + 3;
+    gyro::icntl[Param::DirBC_Interior] = std::get<1>(GetParam());
+    if (gyro::icntl[Param::nr_exp] == 3)
+        gyro::icntl[Param::fac_ani] = 2; //anisotropy should not exceed grid size
+
+    gmgpolar test_p;
+
+    create_grid(test_p);
+
+    level& p_level = *(test_p.v_level[0]);
+    int ctheta_int = test_p.v_level[1]->ntheta_int;
+    int cr_int     = test_p.v_level[1]->nr_int;
+
+    p_level.m  = test_p.v_level[0]->nr * test_p.v_level[0]->ntheta; //number of nodes on fine level
+    p_level.mc = test_p.v_level[1]->nr * test_p.v_level[1]->ntheta; //number of nodes on coarse level
+
+    std::vector<double> u_test(p_level.m);
+    for (int z = 0; z < p_level.m; z++) {
+        u_test[z] = z - 1 + pow(PI, -z * z); //arbitrary grid function to test with
+    }
+
+    std::vector<double> sol = p_level.apply_restriction_inj(u_test);
+
+    EXPECT_EQ((int)sol.size(), p_level.mc);
+
+    for (int j = 0; j < cr_int + 1; j++) {
+        for (int i = 0; i < ctheta_int; i++) {
+            EXPECT_EQ(sol[j * ctheta_int + i], u_test[(2 * j) * p_level.ntheta_int + (2 * i)])
+                << "The injection restriction fails at Index (r,theta): (" + std::to_string(j) + "," +
+                       std::to_string(i) + ")"; //test all values
+        }
+    }
+}
+
+/*!
+ *  \brief Test the extrapolation restriction operator used in the implicit extrapolation step of the multigrid cycle. 
+ *  
+ *  The Test creates an arbitrary grid-function on the finer level and restricts it. 
+ *  On the coarse level we iterate over all nodes and accumulate the adjacent fine node values based on the triangulation.
+ *  We hence only consider 6 adjacent fine nodes, which lie on one of the edges spanned by the corresponding coarse node. 
+ *  This is matrix-free but it corresponds to applying the transposed prolongation matrix
+ *
+ *  Parametrized tests are used to test for different grid sizes and with or without Dirichlet boundary conditions.
+ */
+
+TEST_P(test_restriction, test_extrapolation_restriction)
+{
+    gyro::icntl[Param::nr_exp]         = (int)(std::get<0>(GetParam()) / 3) + 3;
+    gyro::icntl[Param::ntheta_exp]     = (std::get<0>(GetParam()) % 3) + 3;
+    gyro::icntl[Param::DirBC_Interior] = std::get<1>(GetParam());
+    if (gyro::icntl[Param::nr_exp] == 3)
+        gyro::icntl[Param::fac_ani] = 2; //anisotropy should not exceed grid size
+
+    gmgpolar test_p;
+
+    create_grid(test_p);
+
+    level& p_level = *(test_p.v_level[0]);
+    int ctheta_int = test_p.v_level[1]->ntheta_int;
+    int cr_int     = test_p.v_level[1]->nr_int;
+
+    p_level.m  = test_p.v_level[0]->nr * test_p.v_level[0]->ntheta;
+    p_level.mc = test_p.v_level[1]->nr * test_p.v_level[1]->ntheta;
+
+    std::vector<double> u_test(p_level.m);
+    for (int z = 0; z < p_level.m; z++) {
+        u_test[z] = 1 - z; //arbitrary grid function
```

Comment:
why is the arbitrary function changed in every test? Can you make this consistent?

Reply:
done !